File: XMLCheck/Handler/Xml9Parser.cs
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using XMLCheck.Models;

namespace XMLCheck.Handler
{
    public class Xml9Parser : BaseXmlParser, IXmlParser
    {
        public string LoaiHoSo => "XML9";

        public object Parse(string xmlContent)
        {
            var result = new List<Xml9Model>();

            // 1. Load XML (Base xử lý BOM + namespace)
            if (!Load(xmlContent))
                return result;

            // 2. Check root
            if (!IsRoot("CHI_TIEU_DU_LIEU_GIAY_CHUNG_SINH"))
                return result;

            // 3. Lấy danh sách DU_LIEU_GIAY_CHUNG_SINH
            var nodes = Root!
                .Descendants()
                .Where(e => e.Name.LocalName == "DU_LIEU_GIAY_CHUNG_SINH");

            foreach (var n in nodes)
            {
                // ===== Helper cục bộ theo node =====
                string? S9(string k) =>
                    n.Descendants()
                     .FirstOrDefault(x => x.Name.LocalName == k)
                     ?.Value
                     ?.Trim();

                int? I9(string k) =>
                    int.TryParse(S9(k), out var v) ? v : null;

                decimal? D9(string k) =>
                    decimal.TryParse(S9(k), out var v) ? v : null;

                DateTime? DT9(string k)
                {
                    var raw = S9(k);
                    if (string.IsNullOrWhiteSpace(raw)) return null;

                    string[] fmts = { "yyyyMMdd", "yyyyMMddHHmm", "yyyyMMddHHmmss" };
                    return DateTime.TryParseExact(
                        raw,
                        fmts,
                        CultureInfo.InvariantCulture,
                        DateTimeStyles.None,
                        out var d
                    ) ? d : null;
                }

                var m = new Xml9Model
                {
                    MA_LK = S9("MA_LK"),

                    MA_BHXH_NND = S9("MA_BHXH_NND"),
                    MA_THE_NND = S9("MA_THE_NND"),
                    HO_TEN_NND = S9("HO_TEN_NND"),
                    NGAYSINH_NND = DT9("NGAYSINH_NND"),
                    MA_DANTOC_NND = S9("MA_DANTOC_NND"),
                    SO_CCCD_NND = S9("SO_CCCD_NND"),
                    NGAYCAP_CCCD_NND = DT9("NGAYCAP_CCCD_NND"),
                    NOICAP_CCCD_NND = S9("NOICAP_CCCD_NND"),
                    NOI_CU_TRU_NND = S9("NOI_CU_TRU_NND"),

                    MA_QUOCTICH = S9("MA_QUOCTICH"),
                    MATINH_CU_TRU = S9("MATINH_CU_TRU"),
                    MAHUYEN_CU_TRU = S9("MAHUYEN_CU_TRU"),
                    MAXA_CU_TRU = S9("MAXA_CU_TRU"),

                    HO_TEN_CHA = S9("HO_TEN_CHA"),
                    MA_THE_TAM = S9("MA_THE_TAM"),
                    HO_TEN_CON = S9("HO_TEN_CON"),
                    GIOI_TINH_CON = I9("GIOI_TINH_CON"),

                    SO_CON = I9("SO_CON"),
                    LAN_SINH = I9("LAN_SINH"),
                    SO_CON_SONG = I9("SO_CON_SONG"),
                    CAN_NANG_CON = D9("CAN_NANG_CON"),

                    NGAY_SINH_CON = DT9("NGAY_SINH_CON"),
                    NOI_SINH_CON = S9("NOI_SINH_CON"),
                    TINH_TRANG_CON = S9("TINH_TRANG_CON"),

                    SINHCON_PHAUTHUAT = S9("SINHCON_PHAUTHUAT"),
                    SINHCON_DUOI32TUAN = S9("SINHCON_DUOI32TUAN"),

                    GHI_CHU = S9("GHI_CHU"),
                    NGUOI_DO_DE = S9("NGUOI_DO_DE"),
                    NGUOI_GHI_PHIEU = S9("NGUOI_GHI_PHIEU"),
                    NGAY_CT = DT9("NGAY_CT"),

                    SO = S9("SO"),
                    QUYEN_SO = S9("QUYEN_SO"),
                    MA_TTDV = S9("MA_TTDV"),

                    DU_PHONG = S9("DU_PHONG")
                };

                result.Add(m);
            }

            return result;
        }
    }
}

File: XMLValidator_v2/src/parsers/parsers/xml9.parser.ts
import { BaseXmlParser } from '../base-xml.parser';
import { Xml9Model } from '../../interfaces/xml-models.interface';

export class Xml9Parser extends BaseXmlParser {
  public get loaiHoSo(): string {
    return 'XML9';
  }

  public parse(xmlContent: string): Xml9Model[] {
    const result: Xml9Model[] = [];

    if (!this.load(xmlContent)) return result;
    if (!this.isRoot('CHI_TIEU_DU_LIEU_GIAY_CHUNG_SINH')) return result;

    const nodes = this.findDescendants(this.root, 'DU_LIEU_GIAY_CHUNG_SINH');

    for (const n of nodes) {
      const m: Xml9Model = {
        MA_LK: this.getElementText(n, 'MA_LK') || undefined,
        MA_BHXH_NND: this.getElementText(n, 'MA_BHXH_NND') || undefined,
        MA_THE_NND: this.getElementText(n, 'MA_THE_NND') || undefined,
        HO_TEN_NND: this.getElementText(n, 'HO_TEN_NND') || undefined,
        NGAYSINH_NND: this.getElementDate(n, 'NGAYSINH_NND') ?? undefined,
        MA_DANTOC_NND: this.getElementText(n, 'MA_DANTOC_NND') || undefined,
        SO_CCCD_NND: this.getElementText(n, 'SO_CCCD_NND') || undefined,
        NGAYCAP_CCCD_NND: this.getElementDate(n, 'NGAYCAP_CCCD_NND') ?? undefined,
        NOICAP_CCCD_NND: this.getElementText(n, 'NOICAP_CCCD_NND') || undefined,
        NOI_CU_TRU_NND: this.getElementText(n, 'NOI_CU_TRU_NND') || undefined,
        MA_QUOCTICH: this.getElementText(n, 'MA_QUOCTICH') || undefined,
        MATINH_CU_TRU: this.getElementText(n, 'MATINH_CU_TRU') || undefined,
        MAHUYEN_CU_TRU: this.getElementText(n, 'MAHUYEN_CU_TRU') || undefined,
        MAXA_CU_TRU: this.getElementText(n, 'MAXA_CU_TRU') || undefined,
        HO_TEN_CHA: this.getElementText(n, 'HO_TEN_CHA') || undefined,
        MA_THE_TAM: this.getElementText(n, 'MA_THE_TAM') || undefined,
        HO_TEN_CON: this.getElementText(n, 'HO_TEN_CON') || undefined,
        GIOI_TINH_CON: this.getElementInt(n, 'GIOI_TINH_CON') ?? undefined,
        SO_CON: this.getElementInt(n, 'SO_CON') ?? undefined,
        LAN_SINH: this.getElementInt(n, 'LAN_SINH') ?? undefined,
        SO_CON_SONG: this.getElementInt(n, 'SO_CON_SONG') ?? undefined,
        CAN_NANG_CON: this.getElementDecimal(n, 'CAN_NANG_CON') ?? undefined,
        NGAY_SINH_CON: this.getElementDate(n, 'NGAY_SINH_CON') ?? undefined,
        NOI_SINH_CON: this.getElementText(n, 'NOI_SINH_CON') || undefined,
        TINH_TRANG_CON: this.getElementText(n, 'TINH_TRANG_CON') || undefined,
        SINHCON_PHAUTHUAT: this.getElementText(n, 'SINHCON_PHAUTHUAT') || undefined,
        SINHCON_DUOI32TUAN: this.getElementText(n, 'SINHCON_DUOI32TUAN') || undefined,
        GHI_CHU: this.getElementText(n, 'GHI_CHU') || undefined,
        NGUOI_DO_DE: this.getElementText(n, 'NGUOI_DO_DE') || undefined,
        NGUOI_GHI_PHIEU: this.getElementText(n, 'NGUOI_GHI_PHIEU') || undefined,
        NGAY_CT: this.getElementDate(n, 'NGAY_CT') ?? undefined,
        SO: this.getElementText(n, 'SO') || undefined,
        QUYEN_SO: this.getElementText(n, 'QUYEN_SO') || undefined,
        MA_TTDV: this.getElementText(n, 'MA_TTDV') || undefined,
        DU_PHONG: this.getElementText(n, 'DU_PHONG') || undefined,
      };
      result.push(m);
    }

    return result;
  }
}
